lib/auth/auth_with_roles.go

```diff
@@ -1957,6 +2089,46 @@ func (a *ServerWithRoles) generateUserCerts(ctx context.Context, req proto.UserC
 	return certs, nil
 }
 
+// CreateBot creates a new certificate renewal bot and returns a join token.
+func (a *ServerWithRoles) CreateBot(ctx context.Context, req *proto.CreateBotRequest) (*proto.CreateBotResponse, error) {
+	// Note: this creates a role with role impersonation privileges for all
+	// roles listed in the request and doesn't attempt to verify that the
+	// current user has permissions for those embedded roles. We assume that
```

Comment:
Yes, it can. From what I can, tell, though, `VerbCreate` on `Role` is effectively root-level in Teleport today for exactly this reason, so at least I don't think we're introducing a _new_ path to privilege escalation.

We do check for `VerbCreate` on both `User` and `Role` here so I _think_ we have our bases covered but definitely let me know if I've missed something, this is a pretty scary bit of code IMO.

Reply:
@timothyb89 I'm curious if you have considered making a bot a "resource" like any other object in the cluster. Then it could just be managed by usual tctl create/get/rm commands, we could have a `types.KindBot` which would grant users explicit permissions to manage them, and would be easy to reason about "bot" as a "resource" in general. Probably a bit too late to ask design questions at this point but I'm curious if this has ever been considered.

Reply:
That was part of Zac's original design, actually, and I ended up removing it.

Ultimately we needed to reuse so much of Teleport's traditional auth infrastructure (necessitating a backing `User` and `Role`) that the entire `Bot` resource existed solely to be created and removed, alongside the other resources it still needed. Removing the resource was a surprisingly large negative diff and didn't end up breaking any functionality because it just wasn't used.

I wouldn't be surprised to see it come back eventually if we eventually add more of bot-specific metadata. This whole "stuffing bot stuff into `User` labels" strategy is probably not great in the long term .